File: VB/Create-a-Report-Bound-to-JsonDataSource/Form1.vb
﻿Imports DevExpress.DataAccess.Json
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports System.Xml.Linq

Namespace Create_a_Report_Bound_to_JsonDataSource
    Partial Public Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
            Dim report As XtraReport = CreateReport()
            Dim designTool As New ReportDesignTool(report)
            designTool.ShowRibbonDesignerDialog()
        End Sub

        Private Function CreateReport() As XtraReport
            Dim report As New XtraReport()
            Dim DetailBand As New DetailBand()
            DetailBand.HeightF = 50

            Dim XRLabel As New XRLabel()
            XRLabel.WidthF = 300
            XRLabel.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[CompanyName]"))

            DetailBand.Controls.Add(XRLabel)
            report.Bands.Add(DetailBand)



            report.DataSource = CreateDataSourceFromWeb()
            'report.DataSource = CreateDataSourceFromFile();
            'report.DataSource = CreateDataSourceFromText();
            'report.DataMember = "Customers";
            Return report
        End Function
        Private Function CreateDataSourceFromWeb() As JsonDataSource
            Dim jsonDataSource = New JsonDataSource()
            ' Specify a Web Service Endpoint URI with JSON content
            Dim uri = New Uri("http://northwind.servicestack.net/customers.json")
            jsonDataSource.JsonSource = New UriJsonSource(uri)
            Return jsonDataSource
        End Function

        Private Function CreateDataSourceFromFile() As JsonDataSource
            Dim jsonDataSource = New JsonDataSource()
            'Specify the a JSON file's name
            Dim fileUri As New Uri("file:///../../../../customers.txt")
            jsonDataSource.JsonSource = New UriJsonSource(fileUri)
            Return jsonDataSource
        End Function

        Private Function CreateDataSourceFromText() As JsonDataSource

            Dim jsonDataSource = New JsonDataSource()

            'Specify a string with JSON content
            Dim json As String = "{""Customers"":[{""Id"":""ALFKI"",""CompanyName"":""Alfreds Futterkiste"",""ContactName"":""Maria Anders"",""ContactTitle"":""Sales Representative"",""Address"":""Obere Str. 57"",""City"":""Berlin"",""PostalCode"":""12209"",""Country"":""Germany"",""Phone"":""030-0074321"",""Fax"":""030-0076545""}],""ResponseStatus"":{}}"

            ' Specify the object that retrieves JSON data
            jsonDataSource.JsonSource = New CustomJsonSource(json)
            Return jsonDataSource
        End Function
    End Class
End Namespace

File: CS/Create-a-Report-Bound-to-JsonDataSource/Form1.cs
﻿using DevExpress.DataAccess.Json;
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace Create_a_Report_Bound_to_JsonDataSource
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            XtraReport report = CreateReport();
            ReportDesignTool designTool = new ReportDesignTool(report);
            designTool.ShowRibbonDesignerDialog();
        }

        private XtraReport CreateReport()
        {
            XtraReport report = new XtraReport();
            DetailBand DetailBand = new DetailBand();
            DetailBand.HeightF = 50;

            XRLabel XRLabel = new XRLabel();
            XRLabel.WidthF = 300;
            XRLabel.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "[CompanyName]"));

            DetailBand.Controls.Add(XRLabel);
            report.Bands.Add(DetailBand);



            report.DataSource = CreateDataSourceFromWeb();
            //report.DataSource = CreateDataSourceFromFile();
            //report.DataSource = CreateDataSourceFromText();
            //report.DataMember = "Customers";
            return report;
        }
        private JsonDataSource CreateDataSourceFromWeb()
        {
            var jsonDataSource = new JsonDataSource();
            // Specify a Web Service Endpoint URI with JSON content
            var uri = new Uri("http://northwind.servicestack.net/customers.json");
            jsonDataSource.JsonSource = new UriJsonSource(uri);
            return jsonDataSource;
        }

        private JsonDataSource CreateDataSourceFromFile()
        {
            var jsonDataSource = new JsonDataSource();
            //Specify the a JSON file's name
            Uri fileUri = new Uri(@"file:///../../../../customers.txt");
            jsonDataSource.JsonSource = new UriJsonSource(fileUri);
            return jsonDataSource;
        }

        private JsonDataSource CreateDataSourceFromText()
        {

            var jsonDataSource = new JsonDataSource();

            //Specify a string with JSON content
            string json = "{\"Customers\":[{\"Id\":\"ALFKI\",\"CompanyName\":\"Alfreds Futterkiste\",\"ContactName\":\"Maria Anders\",\"ContactTitle\":\"Sales Representative\",\"Address\":\"Obere Str. 57\",\"City\":\"Berlin\",\"PostalCode\":\"12209\",\"Country\":\"Germany\",\"Phone\":\"030-0074321\",\"Fax\":\"030-0076545\"}],\"ResponseStatus\":{}}";

            // Specify the object that retrieves JSON data
            jsonDataSource.JsonSource = new CustomJsonSource(json);
            return jsonDataSource;
        }
    }
}
